lib/cli/src/commands/init.rs

```diff
@@ -0,0 +1,472 @@
+                let _ = std::fs::create_dir_all(s);
+                package_name = s
+                    .canonicalize()?
+                    .file_stem()
+                    .and_then(|s| s.to_str())
+                    .map(|s| s.to_string())
+                    .ok_or_else(|| anyhow::anyhow!("no dir name"))?;
+                s.join(WASMER_TOML_NAME)
+            }
+        };
+
+        if target_file.exists() && !self.overwrite {
+            return Err(anyhow::anyhow!(
+                "wapm project already initialized in {}",
+                target_file.display()
+            ));
+        }
+
+        // See if the directory has a Cargo.toml file, if yes, copy the license / readme, etc.
+        let manifest_path = match self.manifest_path.as_ref() {
+            Some(s) => s.clone(),
+            None => {
+                let cargo_toml_path = self
+                    .out
+                    .clone()
+                    .unwrap_or_else(|| std::env::current_exe().unwrap())
+                    .join("Cargo.toml");
+                cargo_toml_path
+                    .canonicalize()
+                    .unwrap_or_else(|_| cargo_toml_path.clone())
+            }
+        };
+
+        let cargo_toml = if manifest_path.exists() {
+            use anyhow::Context;
+
+            let mut metadata = MetadataCommand::new();
+            metadata.manifest_path(&manifest_path);
+            metadata.no_deps();
+            metadata.features(CargoOpt::AllFeatures);
+
+            println!("{:#?}", metadata.cargo_command());
+
+            let metadata = metadata.exec();
+
+            let metadata = match metadata {
+                Ok(o) => o,
+                Err(e) => {
+                    return Err(anyhow::anyhow!("failed to load metadata: {e}")
+                        .context(anyhow::anyhow!("{}", manifest_path.display())));
+                }
+            };
+
+            let package = metadata
+                .root_package()
+                .ok_or_else(|| anyhow::anyhow!("no root package found in cargo metadata"))
+                .context(anyhow::anyhow!("{}", manifest_path.display()))?;
+
+            Some(MiniCargoTomlPackage {
+                name: package.name.clone(),
+                version: package.version.clone(),
+                description: package.description.clone(),
+                homepage: package.homepage.clone(),
+                repository: package.repository.clone(),
+                license: package.license.clone(),
+                readme: package.readme.clone().map(|s| s.into_std_path_buf()),
+                license_file: package.license_file.clone().map(|f| f.into_std_path_buf()),
+                workspace_root: metadata.workspace_root.into_std_path_buf(),
+                build_dir: metadata
+                    .target_directory
+                    .into_std_path_buf()
+                    .join("wasm32-wasi"),
+            })
+        } else {
+            None
+        };
```

Comment:
Can we pull all of this out into its own `read_cargo_toml()` function? Moving "boring" logic out into its own functions helps with readability because I can skim through `Init::execute()` and see all the interesting bits without having to care about exactly how we get a `MiniCargoTomlPackage`.

Reply:
ok done